Limit log file name length to OS allowed limit

asomsiko opened the pull request.

Commit:
Limit log file name length to OS allowed limit

pbos commented on gtest_parallel.py:
```diff
@@ -44,6 +44,11 @@
   import fcntl
 
 
+# The log file names is truncated to 256 to overcome OS limitations
```
Can you move these constants into where they are used?

asomsiko replied:
Sure. Done

pbos commented on gtest_parallel.py:
```diff
@@ -204,14 +209,18 @@ def _logname(output_dir, test_binary, test_name, execution_number):
+      suffix = '-%d.log' % execution_number
+      prefix = '%s-%s' % (Task._normalize(os.path.basename(test_binary)),
+                          Task._normalize(test_name))
+      namemax = os.statvfs('.').f_namemax if hasattr(os, 'statvfs') else NAMEMAX_DEFAULT
```
Can you have this be: max_prefix_length = be (f_filemax - space_for_random) else default?

asomsiko replied:
Cleaned this up

pbos commented on gtest_parallel.py:
```diff
@@ -204,14 +209,18 @@ def _logname(output_dir, test_binary, test_name, execution_number):
+      prefix = '%s-%s' % (Task._normalize(os.path.basename(test_binary)),
+                          Task._normalize(test_name))
+      namemax = os.statvfs('.').f_namemax if hasattr(os, 'statvfs') else NAMEMAX_DEFAULT
+      (log_handle, log_name) = tempfile.mkstemp(prefix=prefix[:namemax - MKSTEMP_RANDOM_LEN - len(suffix)],
```
Can you rewrite this so that it always tries to create the full file name and then falls back to create a truncated version? That way you can have full-length names even if statvfs isn't supported.

Something like:
* Try creating the file no matter what
* Retry but shortening it using statvfs (or defaults)

asomsiko replied:
Good idea. I've noticed one test insists that the file should be removed after the creation attempt here.

pbos commented on gtest_parallel.py:
```diff
@@ -204,14 +209,18 @@ def _logname(output_dir, test_binary, test_name, execution_number):
+      (log_handle, log_name) = tempfile.mkstemp(prefix=prefix[:namemax - MKSTEMP_RANDOM_LEN - len(suffix)],
+                                                dir=output_dir,
+                                                suffix=suffix)
+    os.close(log_handle)
```
This is a different change from before. Did we have a handle leak here?

asomsiko replied:
There was no handle leak. The `mkstemp` in this change is also used in the second case so `os.close` closes the handle for both

pbos commented on gtest_parallel.py:
```diff
@@ -44,6 +44,11 @@
 
+# The log file names is truncated to 256 to overcome OS limitations
+MKSTEMP_RANDOM_LEN = 12  # length of mkstemp random string being added
+NAMEMAX_DEFAULT = 255  # Limit for log file name length
```
Can this be renamed DEFAULT_PREFIX_LENGTH and maybe = 240?

asomsiko replied:
Sure, renamed and set to 240

Commit:
Attempt creating log file before trimming

Commit:
Add test for log file trimming

Also change root for test_log_file_names tests to be
a temporary directory to allow runnig tests without sudo.
The old root `/` or `C:` is not usually writable for regular users.

Commit:
Cleanup logname stripping comments

pbos commented on gtest_parallel.py:
```diff
@@ -204,14 +204,39 @@ def _logname(output_dir, test_binary, test_name, execution_number):
+        os.remove(log_name)
+    except OSError as os_error:
+      # truncate file name if error is Errno 36: File name too long
+      if os_error.errno == 36:
```
Can this be `errno.ENAMETOOLONG` instead of the comment?

pbos replied:
https://docs.python.org/3/library/errno.html#errno.ENAMETOOLONG

pbos commented on gtest_parallel.py:
```diff
@@ -204,14 +204,39 @@ def _logname(output_dir, test_binary, test_name, execution_number):
-                                 Task._normalize(test_name), execution_number)
+    # The log file name is a combination of a test binary name and a test name.
+    # If the log file name exceeds OS limitations it will be truncated.
+    MKSTEMP_RANDOM_LEN = 12  # length of mkstemp random string being added
```
"# Space reserved for mkstemp suffix."

pbos commented on gtest_parallel.py:
```diff
@@ -204,14 +204,39 @@ def _logname(output_dir, test_binary, test_name, execution_number):
+      # truncate file name if error is Errno 36: File name too long
+      if os_error.errno == 36:
+        prefix_length = ((os.statvfs('.').f_namemax -
+                          MKSTEMP_RANDOM_LEN) if hasattr(os, 'statvfs') else MAX_PREFIX_LENGTH) - len(suffix)
```
I think this would read easier as: if hasattr(..): prefix_length = ..., i.e. avoid the ternary. as it's long.

pbos commented on gtest_parallel_tests.py:
```diff
@@ -311,36 +311,44 @@ def test_times_worker():
+                                    'Test.case', 1))
+      MAX_PREFIX_LENGTH = 240
+      long_test_name = 'a' * (os.statvfs('.').f_namemax if hasattr(os,
+                                                                   'statvfs') else MAX_PREFIX_LENGTH + 1)
```
Could you make this one a separate test? Maybe `test_truncates_file_name`

Also I think this test only makes sense on systems that have `statvfs`, can it run conditionally? This test requires OSes without `statvfs` to reject filenames that are longer which is maybe not what we want. I'm OK with not testing the MAX_PREFIX_LENGTH branch.

pbos reviewed (commented):
Sorry for the slow response time, new parent here.
